### ISSUE TITLE
Use pkgconf as our pkg-config implementation

### ISSUE THREAD

**Issue by emilazy** [NixOS/nixpkgs#334195]
The original [freedesktop pkg-config implementation](https://freedesktop.org/wiki/Software/pkg-config/) we use is pretty dead (no stable release since 2017, no commits in 3 years, no code changes since 2019), and has a variety of issues, including an awkward circular GLib dependency that makes bootstrapping more annoying, and missing support for useful `.pc` features.

pkgconf is a compatible from‐scratch reimplementation that avoids all these issues; they have a (biased, naturally, but still) [comparison to pkg-config](http://pkgconf.org/features.html). Fedora has already [moved to pkgconf](https://fedoraproject.org/wiki/Changes/pkgconf_as_system_pkg-config_implementation) as its pkg-config implementation, as have Arch, Alpine, and [Debian](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1020799).

@reckenrode will be using pkgconf in the Darwin SDK rework because of build issues with pkg-config’s vendored GLib, so the problems here are not just theoretical. I think we should strongly consider moving wholesale to pkgconf.

**Comment by reckenrode** [comment]
> @reckenrode will be using pkgconf in the Darwin SDK rework because of build issues with pkg-config’s vendored GLib, so the problems here are not just theoretical. I think we should strongly consider moving wholesale to pkgconf.

One thing I’d like to clarify is that it’s used only in the bootstrap. The final package set will have whatever pkg-config implementation is used by default in nixpkgs.

The problem for Darwin is that pkg-config bundles its own glib, which fails to build in the new bootstrap because it needs `arpa/nameser.h`, which is provided by libresolv, which is built with Meson, which depends on pkg-config. I need to break the chain of circular dependencies somewhere, which I have opted to do by making pkgconf the top-level pkg-config in the bootstrap.

(For those who are not aware, it’s such a hassle dealing with xcbuild to build source release packages that it’s easier/faster to reimplement their build systems using Meson. I chose Meson because it is easy to use and readable.)

**Comment by emilazy** [comment]
Right, I didn’t mean to imply that we’re getting a platform split on this, just to point to an example of how our existing pkg-config implementation is actively worse for us in practice such that we resort to pkgconf anyway. pkg-config’s bootstrapping issues are well‐known in other distros and a big motivator for why the ecosystem has largely moved over to pkgconf.

**Comment by reckenrode** [comment]
Sure, sorry if I came across like you were. I wanted to address those less familiar with what’s in the plans for Darwin in case someone got an incorrect impression about it.

**Comment by RossComputerGuy** [comment]
> @reckenrode will be using pkgconf in the Darwin SDK rework because of build issues with pkg-config’s vendored GLib

I'm looking to have an LLVM based bootstrap in the future possibly and from the looks of it, we might run into the same exact problem as Darwin because of this problem. Using pkgconfig instead of pkg-config sounds great.

**Comment by emilazy** [comment]
Working on this.

**Comment by theoparis** [comment]
I would also like to see this happen. I also noticed that the `pkgconf` derivation doesn't have the `pkg-config` binary at the moment.

**Comment by tobim** [comment]
There is an important difference between the two in `nixpkgs` tho:
`pkg-config` is patched to ignore `Requires.private` for `--cflags` without `--static`, whereas `libpkgconf` is not. I believe such a change also has to be made for `pkgconf` to be viable.

**Comment by emilazy** [comment]
That patch is bad and has caused us problems. Avoiding it would be one of the benefits of switching we discussed when running into issues with `wlroots` recently because of it. We would have to either propagate more packages or (preferably) figure out a way to have absolute paths to private `.pc` dependencies.

**Comment by N-R-K** [comment]
> `pkg-config` is patched to ignore `Requires.private` for `--cflags` without `--static`, whereas `libpkgconf` is not.

Okay so I've been looking into this a bit and here are my findings. First a real world example using `xft`:

```
[/tmp]~> cat test.c
#include <X11/Xft/Xft.h>
int main(void) { (void)&XftFontClose; return 0; }
[/tmp]~> grep 'Requires' /usr/lib64/pkgconfig/xft.pc
Requires: xproto
Requires.private: xrender, fontconfig, freetype2
[/tmp]~> pkg-config --cflags xft
-I/usr/include/freetype2 -I/usr/include/harfbuzz -I/usr/include/glib-2.0 -I/usr/lib64/glib-2.0/include
[/tmp]~> gcc -o /dev/null test.c $(pkg-config --cflags --libs xft)
```

Using the "standard" pkg-config behavior, it all goes well. Now if I use a patched version (https://github.com/skeeto/u-config/commit/62c88c0803f05ba33deb6151a20ca5d6b61598b6) that ignores `Requires.private` for non static invocation, I get an empty `--cflags` and the build fails:

```
[/tmp]~> ./pkg-config-ignore-private-cflags --cflags xft

[/tmp]~> gcc -o /dev/null test.c $(./pkg-config-ignore-private-cflags --cflags --libs xft)
In file included from test.c:1:
/usr/include/X11/Xft/Xft.h:40:10: fatal error: ft2build.h: No such file or directory
   40 | #include <ft2build.h>
      |          ^~~~~~~~~~~~
compilation terminated.
```

The is because xft includes freetype headers - which the compiler fails to find without freetype's `Cflags:`.

- - -

In short, there's 3 cases:

1) Required, always (`Requires:`).
2) Required, only when static.
3) Required when static, but headers (`Cflags:`) required always. (`Requires.private`).

nix seems to be assuming the 2nd kind for transitive dependencies (which IMO is reasonable), but `pkg-config` assumes the 3rd kind. Worse, pc files seems to have no way of differentiating between the 2nd and 3rd case. And so `pkg-config` treats everything as the 3rd case, even when it might've been the 2nd case to be "safe".

I don't know what should or could be done on nix's side, but my initial impression is that ideally pc files should allow some way (`Requires.really_private` or w/e) for packages to describe themselves as the 2nd kind rather than implicitly degrading to the 3rd.

(ping: @skeeto @marcin-serwin)

**Comment by tobim** [comment]
The analysis seems right. To my understanding "standard" `pkg-config` just assumes that header declarations of a transitive dependency are automatically part of the direct dependencies API surface. This is essentially case (3) from your interpretation. I actually don't understand what kind of dependency relationship should be modeled with this, because when you include a header from another lib, it should absolutely be `Required` unconditionally. 

That all means `*.pc` files introduce false interface dependencies all over the place because their authors can only use `Requires.private` when they really need something for case (2).

The solution taken in nixpkgs is to change (3) into (2) and model the actually correct dependency graph with `propagated*Inputs` with the help of `cc-wrapper`. That way `-I` and `-isystem` flags are taken care of independent from `pkg-config`. Any other flags that show up in transitive `Cflags:` aren't covered by that, but in my impression that doesn't introduce a lot of problems.

**Comment by emilazy** [comment]
FWIW, I remember propagating the additional dependencies seeming like it would be viable and not *too* much of a headache, just not as clean as absolute path references if we could achieve them. But yes, the patch we have changes semantics in ways that break the expectations of `.pc` authors and have led to actual confusion/breakage in Nixpkgs, so it’s not a good thing, even if the upstream semantics are bad.

**Comment by emilazy** [comment]
Do they actually? I remember source-diving into `pkgconf` and it didn’t seem to. But that’d be a very nice surprise if so.

**Comment by emilazy** [comment]
Oh, wonderful! Then yeah we can just have a hook to rewrite everything appropriately and the problem is solved.

**Comment by RossSmyth** [comment]
I've found an issue where using pkgconf (which we use at work by default), udev is unable to be found. But switching to pkg-config it works fine. For example, a build failed for a Rust-based tool, and this is the patch that fixed it:
```patch
--- a/nix/default.nix
+++ b/nix/default.nix
@@ -3,7 +3,7 @@
   stdenv,
   craneLib,
   udev,
-  pkgconf,
+  pkg-config,
 }:
 let
   fs = lib.fileset;
@@ -27,7 +27,7 @@
     strictDeps = true;
 
     nativeBuildInputs = lib.optionals stdenv.hostPlatform.isUnix [
-      pkgconf
+      pkg-config
     ];
 
     buildInputs = lib.optionals stdenv.hostPlatform.isUnix [
```

The only pkg-config-based dep is udev.
